functorch/_src/aot_autograd.py

```diff
@@ -394,34 +400,31 @@ def create_aot_dispatcher_function(
     The resulting compiled forward and backward graphs are then wrapped up in a
     ``torch.autograd.Function`` object.
     """
+
+    # This is the main entry point.
+    # TODO: Chillee argues that dynamo itself should pass in fake tensors to
+    # the list of arguments when compiling; at the moment we do not do this
+
     if aot_config.decompositions is None:
         aot_config.decompositions = {}
 
     aot_config.decompositions = {
         **aot_autograd_decompositions,
         **aot_config.decompositions,
     }
-    fake_mode = FakeTensorMode if config.use_fake_tensor else nullcontext
+    # NB: don't bother setting allow_fallback_kernels; this should not actually
+    # be configurable in fake tensor, we should automatically do the right
+    # thing
+    fake_mode = FakeTensorMode() if config.use_fake_tensor else nullcontext()
+    python_dispatcher_mode = enable_python_dispatcher() if config.use_dynamic_shapes else nullcontext()
+    shape_env = ShapeEnv() if config.use_dynamic_shapes else None
 
-    with preserve_rng_state(), fake_mode() as mode:
+    with preserve_rng_state(), fake_mode, python_dispatcher_mode:
 
         def process_inputs(flat_args):
-            if mode:
-                seen_args = set()
-
+            if config.use_fake_tensor:
                 def convert(x):
-                    # HACK HACK HACK
-                    # preserve the same behavior of the non-fake tensor branch
-                    # of creating a unique tensor impl for each input,
-                    # instead of memoizing the conversion. this has the same
-                    # problem of models that resize their inputs described below,
-                    # but fixes an issue with tied parameters.
-                    # TODO: more full fix
-                    if id(x) in seen_args:
-                        with torch.utils._mode_utils.no_dispatch():
-                            x = x.detach().requires_grad_(x.requires_grad)
-                    seen_args.add(id(x))
-                    return mode.from_tensor(x)
+                    return fake_mode.from_tensor(x, shape_env=shape_env)
```

Comment:
I couldnt figure out a good small representative test for this but without this fix like half of the bert models break, I dont think we can just get rid of this logic atm

Reply:
@eellison that doesn't make sense, I have tested BERT_pytorch with this change and it passes... which models?

Reply:
`python benchmarks/huggingface.py --ci -d cuda --float32 --backend=aot_nop --training --only=RobertaForCausalLM`

Reply:
https://github.com/pytorch/pytorch/issues/85093

Reply:
works on the symbolic-shapes branch (modulo accuracy problems lol)

```
torchdynamo.utils: [WARNING] RMSE (res-fp64): 0.00034, (ref-fp64): 0.00016
INCORRECT
Command exited with non-zero status 255
17.88user 18.49system 0:19.60elapsed 185%CPU (0avgtext+0avgdata 2531928maxresident)k
0inputs+296outputs (0major+645318minor)pagefaults 0swaps
[ezyang@devgpu001.pnb3 ~/local/pytorch-tmp (symbolic-shapes)]$ (cd ../torchdynamo && pp TORCH_SHOW_CPP_STACKTRACES=1 AOT_FX_GRAPHS_JOINT=1 TORCHDYNAMO_DYNAMIC_SHAPES=0 AOT_DYNAMIC_SHAPES=0 time python benchmarks/huggingface.py --only RobertaForCausalLM --accuracy-aot-nop --training) 
```